From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SkiffOS Contributors <skiff@skiffos.org>
Date: Wed, 02 Apr 2025 00:00:00 +0000
Subject: [PATCH] ARM: dts: sun8i: add Clockwork Pi CPI3 (GameShell)

Add device tree support for the Clockwork Pi CPI3, sold as the
GameShell handheld game console. The board is based on the Allwinner
R16/A33 SoC (sun8i-a33) with the following peripherals:

  - AXP223 PMIC on R_RSB bus (address 0x3a3)
  - Broadcom BCM43xx WiFi+BT combo module on MMC1 (SDIO, PG0-PG5)
    WiFi reset: PL6, BT reset: PL8, BT shutdown: PL10
  - KD027 2.7" 320x240 display wired via GPIO bit-banged SPI:
    PC0=MOSI, PC2=SCK, PC3=CS, PB2=RST, PH1=backlight (gpio-backlight)
  - USB OTG (MUSB), ID detect on PL4, VBUS drive via AXP drivevbus
  - UART0 console on PB8/PB9 (uart0_pb_pins)
  - MicroSD on MMC0 (PF0-PF5), card-detect on PB3
  - Speaker amplifier enable on PL3 (in codec_analog node)
  - GPIO LED on PB7 (active-low)

Hardware wiring verified against the vendor Linux 5.15 patches:
  https://github.com/clockworkpi/GameShell/tree/main/Code/Kernel/v0.6

Signed-off-by: SkiffOS Contributors <skiff@skiffos.org>
---
 arch/arm/boot/dts/allwinner/Makefile                       |   1 +
 arch/arm/boot/dts/allwinner/sun8i-r16-clockworkpi-cpi3.dts | 259 +++++++++++++++++++++++++++
 2 files changed, 260 insertions(+)
 create mode 100644 arch/arm/boot/dts/allwinner/sun8i-r16-clockworkpi-cpi3.dts

diff --git a/arch/arm/boot/dts/allwinner/Makefile b/arch/arm/boot/dts/allwinner/Makefile
index 0000000..0000001 100644
--- a/arch/arm/boot/dts/allwinner/Makefile
+++ b/arch/arm/boot/dts/allwinner/Makefile
@@ -1,6 +1,7 @@
 	sun8i-r16-bananapi-m2m.dtb \
 	sun8i-r16-nintendo-nes-classic.dtb \
 	sun8i-r16-nintendo-super-nes-classic.dtb \
+	sun8i-r16-clockworkpi-cpi3.dtb \
 	sun8i-r16-parrot.dtb \

diff --git a/arch/arm/boot/dts/allwinner/sun8i-r16-clockworkpi-cpi3.dts b/arch/arm/boot/dts/allwinner/sun8i-r16-clockworkpi-cpi3.dts
new file mode 100644
index 0000000..0000001
--- /dev/null
+++ b/arch/arm/boot/dts/allwinner/sun8i-r16-clockworkpi-cpi3.dts
@@ -0,0 +1,259 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) ClockworkPi developers
+ * Copyright (C) 2025 SkiffOS contributors
+ *
+ * Device tree for the Clockwork Pi CPI3 (GameShell).
+ * Hardware wiring based on the vendor Linux 5.15 patches (v0.6).
+ */
+
+/dts-v1/;
+#include "sun8i-a33.dtsi"
+#include "sunxi-common-regulators.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+
+/ {
+	model = "Clockwork CPI3";
+	compatible = "clockwork,clockworkpi-cpi3", "allwinner,sun8i-r16",
+		     "allwinner,sun8i-a33";
+
+	aliases {
+		serial0 = &uart0;
+		mmc0 = &mmc0;
+		mmc1 = &mmc1;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	/*
+	 * OCP8178 GPIO backlight on PH1.
+	 */
+	backlight: backlight {
+		compatible = "gpio-backlight";
+		gpios = <&pio 7 1 GPIO_ACTIVE_HIGH>; /* PH1 */
+		default-on;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-0 {
+			label = "clockworkpi:green:usr";
+			gpios = <&pio 1 7 GPIO_ACTIVE_LOW>; /* PB7 */
+			linux,default-trigger = "heartbeat";
+		};
+	};
+
+	/*
+	 * Broadcom BCM43xx WiFi+BT power sequence.
+	 * WiFi reset on PL6 (active-low).
+	 */
+	wifi_pwrseq: wifi-pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&r_pio 0 6 GPIO_ACTIVE_LOW>; /* PL6 */
+		post-power-on-delay-ms = <200>;
+	};
+
+	/*
+	 * Bluetooth rfkill control.
+	 * Reset on PL8, shutdown on PL10.
+	 */
+	rfkill_bt: rfkill-bt {
+		compatible = "rfkill-gpio";
+		rfkill-name = "bluetooth";
+		rfkill-type = "bluetooth";
+		reset-gpio = <&r_pio 0 8 GPIO_ACTIVE_HIGH>;  /* PL8 */
+		shutdown-gpio = <&r_pio 0 10 GPIO_ACTIVE_LOW>; /* PL10 */
+	};
+
+	/*
+	 * KD027 2.7" 320x240 display on GPIO bit-banged SPI.
+	 *
+	 * The KD027 uses a custom 9-bit GPIO SPI protocol without a
+	 * separate DC/RS pin. The 'clockwork,cpi3-lcd' compatible
+	 * requires the out-of-tree kd027-lcd driver. The fallback
+	 * 'ilitek,ili9341' compatible activates the fbtft staging driver
+	 * for basic framebuffer output.
+	 */
+	spi_lcd: spi-gpio {
+		compatible = "spi-gpio";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		sck-gpios  = <&pio 2 2 GPIO_ACTIVE_HIGH>; /* PC2 */
+		mosi-gpios = <&pio 2 0 GPIO_ACTIVE_HIGH>; /* PC0 */
+		cs-gpios   = <&pio 2 3 GPIO_ACTIVE_HIGH>; /* PC3 */
+		num-chipselects = <1>;
+
+		panel@0 {
+			compatible = "ilitek,ili9341";
+			reg = <0>;
+			spi-max-frequency = <3200000>;
+			spi-3wire;
+			reset-gpios = <&pio 1 2 GPIO_ACTIVE_LOW>; /* PB2 */
+			backlight = <&backlight>;
+			rotate = <270>;
+			bgr;
+			fps = <30>;
+			buswidth = <9>;
+			debug = <0>;
+		};
+	};
+};
+
+&codec {
+	status = "okay";
+};
+
+/*
+ * codec_analog: sun8i-a33 analog codec.
+ * speaker-amplifier-gpios controls the external amp enable on PL3.
+ * interrupts line matches the codec_analog IRQ in sun8i-a33.dtsi.
+ */
+&codec_analog {
+	pinctrl-names = "default";
+	pinctrl-0 = <&speaker_amplifier_pin>;
+	speaker-amplifier-gpios = <&r_pio 0 3 GPIO_ACTIVE_HIGH>; /* PL3 */
+	cpvdd-supply = <&reg_aldo1>;
+	interrupts = <GIC_SPI 46 IRQ_TYPE_LEVEL_HIGH>;
+	status = "okay";
+};
+
+&cpu0 {
+	cpu-supply = <&reg_dcdc3>;
+};
+
+&de {
+	status = "okay";
+};
+
+&ehci0 {
+	status = "okay";
+};
+
+&i2c0 {
+	status = "okay";
+};
+
+/*
+ * MMC0: MicroSD card slot on PF0-PF5.
+ * Card detect on PB3 (active-low) via mmc0_cd_pin pinctrl group.
+ */
+&mmc0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc0_pins>, <&mmc0_cd_pin>;
+	vmmc-supply = <&reg_dcdc1>;
+	cd-gpios = <&pio 1 3 GPIO_ACTIVE_LOW>; /* PB3 */
+	bus-width = <4>;
+	status = "okay";
+};
+
+/*
+ * MMC1: Broadcom BCM43xx WiFi on SDIO (PG0-PG5).
+ * Non-removable; power sequenced via wifi_pwrseq.
+ */
+&mmc1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&mmc1_pg_pins>, <&wifi_reset_pin>;
+	vmmc-supply = <&reg_aldo1>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <4>;
+	non-removable;
+	keep-power-in-suspend;
+	status = "okay";
+};
+
+&ohci0 {
+	status = "okay";
+};
+
+/*
+ * AXP223 PMIC on R_RSB (Reduced Serial Bus), address 0x3a3.
+ */
+&r_rsb {
+	status = "okay";
+
+	axp223: pmic@3a3 {
+		compatible = "x-powers,axp223";
+		reg = <0x3a3>;
+		interrupt-parent = <&nmi_intc>;
+		interrupts = <0 IRQ_TYPE_LEVEL_LOW>;
+		drivevbus-supply = <&reg_vcc5v0>;
+		x-powers,drive-vbus-en;
+
+		regulators {
+			reg_aldo1: aldo1 {
+				regulator-name = "vcc-io";
+				regulator-always-on;
+				regulator-min-microvolt = <3000000>;
+				regulator-max-microvolt = <3000000>;
+			};
+
+			reg_aldo2: aldo2 {
+				regulator-name = "vdd-dll";
+				regulator-always-on;
+				regulator-min-microvolt = <2500000>;
+				regulator-max-microvolt = <2500000>;
+			};
+
+			reg_aldo3: aldo3 {
+				regulator-name = "vcc-pll-avcc";
+				regulator-always-on;
+				regulator-min-microvolt = <3000000>;
+				regulator-max-microvolt = <3000000>;
+			};
+
+			reg_dc5ldo: dc5ldo {
+				regulator-name = "vdd-cpus";
+				regulator-always-on;
+				regulator-min-microvolt = <1100000>;
+				regulator-max-microvolt = <1100000>;
+			};
+
+			reg_dcdc1: dcdc1 {
+				regulator-name = "vcc-3v0";
+				regulator-always-on;
+				regulator-min-microvolt = <3000000>;
+				regulator-max-microvolt = <3000000>;
+			};
+
+			reg_dcdc2: dcdc2 {
+				regulator-name = "vdd-sys";
+				regulator-always-on;
+				regulator-min-microvolt = <1100000>;
+				regulator-max-microvolt = <1100000>;
+			};
+
+			reg_dcdc3: dcdc3 {
+				regulator-name = "vdd-cpu";
+				regulator-min-microvolt = <900000>;
+				regulator-max-microvolt = <1400000>;
+			};
+
+			reg_dcdc5: dcdc5 {
+				regulator-name = "vcc-dram";
+				regulator-always-on;
+				regulator-min-microvolt = <1500000>;
+				regulator-max-microvolt = <1500000>;
+			};
+
+			reg_dldo1: dldo1 {
+				regulator-name = "vcc-wifi0";
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+			};
+
+			reg_dldo2: dldo2 {
+				regulator-name = "vcc-wifi1";
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+			};
+
+			reg_dldo3: dldo3 {
+				regulator-name = "vcc-amp";
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+			};
+
+			reg_dldo4: dldo4 {
+				regulator-name = "vcc-wifi2";
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+			};
+
+			reg_drivevbus: drivevbus {
+				regulator-name = "usb0-vbus";
+				status = "okay";
+			};
+
+			reg_rtc_ldo: rtc-ldo {
+				regulator-name = "vcc-rtc";
+				regulator-always-on;
+				regulator-min-microvolt = <3000000>;
+				regulator-max-microvolt = <3000000>;
+			};
+		};
+	};
+};
+
+/* UART0 console on PB8/PB9 */
+&uart0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart0_pb_pins>;
+	status = "okay";
+};
+
+/* USB OTG; ID detect on PL4, VBUS supplied via AXP223 drivevbus */
+&usb_otg {
+	dr_mode = "otg";
+	status = "okay";
+};
+
+&usbphy {
+	usb0_id_det-gpios = <&r_pio 0 4 GPIO_ACTIVE_HIGH>; /* PL4 */
+	usb0_vbus-supply = <&reg_drivevbus>;
+	usb1_vbus-supply = <&reg_dldo1>;
+	status = "okay";
+};
